### PR TITLE
Resolving signature/descriptor missmatch

### PR DESCRIPTION
**Pull request by majakusber**
The problem here was the following:
- signature, if it exists, contains information about generics,
- descriptor always exists but does not contain generics info.
Therefore, when possible, we want to use signature. In some cases, such as constructors of enums and certain inner classes, the descriptor may contain higher (correct) number of parameters than the signature, so we want to use descriptor here (at the cost of losing generic information, for now).

### CODE REVIEW

**Comment by majakusber** [comment]
@thk123 @chrisr-diffblue  Could you please review?

**Comment by chrisr-diffblue** [comment]
A small general comment (nit really): Could you give PR's a more meaningful name? It makes it much nicer when skimming through the commit/merge history to see what has been merged in without having to look it up in Jira. I don't know if you can change the title of a PR once it's created, but I assume there is some way :-)

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Nit: `loosing`->`losing`

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (chrisr-diffblue) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Is there any mileage in slightly restructuring the code so that you always initialise `member_type` to `java_type_from_string(m.descriptor)`, then do your `if(m.signature.has_value())` and `try` block, then your `catch` block would just be empty (probably with a comment to explain) and you could lose the final `else` block? That potentially adds a cost in wasted work, but loses some duplicate code and also loses the (depending on who you ask) questionable use of exceptions for control flow :-) @thk123 might have a better view than me here though.

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp:32]: Nice :-) I should copy  that pattern for my unit tests...

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp]: Is this still mean to be here, or is this left over debug?

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp]: Is this include just left over debug?

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (majakusber) [reply unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp]: Leftover debug, good catch. Will do a quick fix.

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp]: Nit: Either break the line after the `=` (my personal favourite) or after the `(`

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (majakusber) [reply unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp]: Will check, probably it is, connected to the symbol table show.

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp]: Nit(s): No spaces either side of the `=`, and break after the `(`. I might be tempted to actually break after `=` if that meant the rest of the expression would fit on a single line.

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp]: Nit : break after the `(`

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp]: This pattern of `REQUIRE(ID_struct)`, `REQUIRE(is_class)` appears a couple of times - it could be factored out into a utility function in `testing-utils` - e.g. see `require_expr.h`. I have a `require_types.h` in there as part of PR1460, if you want an example

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp]: Nit: CLion default line breaking strikes again... The indent should be 2 spaces.

File: unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp
Comment (chrisr-diffblue) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp]: Nit: Break after the `(`, and no spaces around the `=`. See also my comment further back that you may want to break after the `=` instead.

**Review by chrisr-diffblue — CHANGES_REQUESTED** [review]
Needs the formatting cleaned up, but looks pretty good otherwise.

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (chrisr-diffblue) [reply src/java_bytecode/java_bytecode_convert_method.cpp]: Another small comment having thought about this a bit overnight - whether you restructure the code or not, could you add a comment prior to, or inside, this block of code to explain the rationale for having this code? While we currently have an understanding of why we have to do this dance with the signature/descriptor, in a few weeks/months time it might not be so fresh in our memories, or if teams shuffle around and someone new to the generics stuff comes across this code, the first question they'll probably have is 'Why are we doing this?!?' :-)

Comment (thk123) [reply src/java_bytecode/java_bytecode_convert_method.cpp]: Probably the easiest way to deal with both these comments is to pull it out into a function (you can add a comment on the function and directly assign the `member_type`. I think doing the descriptor up front is a reasonable suggestion as is only wasted work when there is no signature which actually based on what Matthias said can't happen for methods even without debug symbols.

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: I suspect we should log a warning whenever this happens saying `"could not parse signature for method_identifier, falling back to descriptor: e.message"`

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Nit: prefer full variable names (e.g. `member_type_from_signature`, `member_type_from_descriptor`)

**Review by thk123 — APPROVED** [review]
Some suggestions, I also agree with what Chris has commented with, but nothing blocking on top of that

**Comment by majakusber** [comment]
Thanks for comments, all should be taken care of now, could you please re-review? @chrisr-diffblue especially take a look at testing-utils/require_class_not_stub please

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Sorry, could this be documented using a doxygen style (inserting `///` and pressing enter in CLion will auto-generate a stub to fill in. This allows it to end up in the auto documentation.

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (chrisr-diffblue) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Can't you do `messaget message(message_handler)` here? (I've currently got an out of date branch checked out so I might be looking at an old constructor here.

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (chrisr-diffblue) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: Nit - maybe avoid breaking before `<< message.eom` ?

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (thk123) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: You might consider outputting the descriptor (or/and signature?) here to help track down what isn't being parsed

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (chrisr-diffblue) [inline src/java_bytecode/java_bytecode_convert_method.cpp:328]: Please update the Doxygen comments for this function to reflect the extra parameter.

File: src/java_bytecode/java_bytecode_convert_method.h
Comment (chrisr-diffblue) [inline src/java_bytecode/java_bytecode_convert_method.h:57]: Nit: Even though it's not strictly necessary, I tend to prefer naming the parameters in the declaration - e.g. `message_handlert &message_handler` - but I note that the `java_bytecode_parse_treet::methodt` parameter hasn't been named either (I wouldn't be against you doing a drive-by cleanup of that at the same time ;-) )

File: unit/testing-utils/require_class_not_stub.cpp
Comment (chrisr-diffblue) [inline unit/testing-utils/require_class_not_stub.cpp]: If it's not too much of a pain, I'd be tempt to suggest you name this file (and it's corresponding header file, of course) to something more generic - e.g. `require_symbol.cpp/.h` so that in future other helper functions can be added. I suggest `require_symbol` only because the one utility function it currently contains takes a `symbolt` as an argument, but feel free to choose another name if you feel its more appropriate.

File: unit/testing-utils/require_class_not_stub.cpp
Comment (chrisr-diffblue) [inline unit/testing-utils/require_class_not_stub.cpp]: Is a more descriptive name for this function perhaps something like `require_complete_class` ? It's not immediately obvious to me what "not_stub" means.

File: unit/testing-utils/require_class_not_stub.cpp
Comment (chrisr-diffblue) [reply unit/testing-utils/require_class_not_stub.cpp]: Oh, and if you do rename the file, remember to update the #ifdef's in the header file to match the new file name...

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (chrisr-diffblue) [inline src/java_bytecode/java_bytecode_convert_method.cpp]: I think you can make the `message_handler` parameter be `const`?

**Review by chrisr-diffblue — APPROVED** [review]
Looks good, nothing blocking, but a few small suggestions for you to consider.

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (majakusber) [reply src/java_bytecode/java_bytecode_convert_method.cpp]: You're right, that's the correct constructor, I will correct this. In order to use it, I cannot make the `const` though.

File: src/java_bytecode/java_bytecode_convert_method.h
Comment (majakusber) [reply src/java_bytecode/java_bytecode_convert_method.h:57]: I asked about the general guidelines for this recently and I was told that the names can be left out unless it's not clear from the type what the parameter represents. For example, from `const irep_idt &` it would not be clear that the parameter is the method identifier, thus the name is given. But for `const java_bytecode_parse_treet::methodt &` it's pretty clear what it is. Since the message handler does not have any 'hidden' role here, I would probably leave it.

File: src/java_bytecode/java_bytecode_convert_method.cpp
Comment (chrisr-diffblue) [reply src/java_bytecode/java_bytecode_convert_method.cpp]: Ah, as @svorenova pointed out to me, `message_handler` can't actually be `const`, so this comment is garbage...

File: src/java_bytecode/java_bytecode_convert_method.h
Comment (chrisr-diffblue) [reply src/java_bytecode/java_bytecode_convert_method.h:57]: Fine by me then :-)

**Review by thk123 — CHANGES_REQUESTED** [review]
Again I highly agree with Chris's comments - please fix the doxygen stuff I pointed out as well then I think this should be good to go (pending getting the TG bump passing)

File: unit/testing-utils/require_symbol.h
Comment (chrisr-diffblue) [inline unit/testing-utils/require_symbol.h]: Nit - it's usually nice with `#endif`s to include the DEFINE that they are ending as a comment at the end of the line, e.g. 
```
#endif // CPROVER_TESTING_UTILS_REQUIRE_SYMBOL_H
```
You'll see this pattern in other header files. While it's not a big deal with a header file as small as this, it's a good habit to get into because if you ever end up working on a file with multiple nested `#ifdef`s, it really helps to disambiguate the endings :-)

**Review by chrisr-diffblue — APPROVED** [review]
Looks good to go if you can just fix up the `#endif` nit I mentioned.

**Review by thk123 — APPROVED** [review]

**Comment by thk123** [comment]
For some reason it is saying the doxygen check failed (meaning the doxygen is not well formed or contains errors like missing parameters from documentation) but the log doesn't say why. You can run it locally using:

```
scripts/run_diff.sh DOXYGEN upstream/develop
```

and see if you get any errors.

**Review by thk123 — CHANGES_REQUESTED** [review]
Blocking so we remember to update these tests once they are merged in: #1494

**Comment by thk123** [comment]
#1494 is now merged so if you rebase you should be able to update the unit test appropriately.

**Comment by majakusber** [comment]
Great thank you, on it..

File: unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp
Comment (thk123) [inline unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp:177]: Above comment can be removed.

**Review by thk123 — APPROVED** [review]
Pending CI (and comment removed please :)

File: unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp
Comment (thk123) [reply unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp:177]: Incorrect - comment relates to a different issue

**Pull request merged**
